### ISSUE TITLE
Support for azurerm_data_factory_pipeline to allow for parameters and variables of non string types

### ISSUE THREAD

**Issue by johndvnguyen** [hashicorp/terraform-provider-azurerm#14198]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Datafactory pipelines created via terraform have the optional arguments of parameters and variables.
These arguments only accept string maps of the parameter/variable names at this time. 

It would great if these accepted parameter/variable blocks that allowed types similar to how global variables for the datafactory are created

### New or Affected Resource(s)

* azurerm_data_factory_pipeline

### Potential Terraform Configuration

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.
```

### References

* #0000

**Comment by rbollet** [comment]
Hello
I have same problem, I need to deploy an array variable type.
Do you have some update ?

**Comment by embarrassedmilk** [comment]
As far as I understand, it was implemented like this due to the limitations in Terraform SDK (correct me if I'm wrong). Now I'm curious what was the reason to choose a map of strings instead of accepting plain json (like in `activities_json`).

If you guys would be open for a PR that adds `variables_json`, I could give it a go

**Comment by gkman** [comment]
I do believe I looked at this lat week and saw that the AzureSDK supports other options, which is what the provider is using under the hood. If you follow the trace

* [Expansion of parameters & variables](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/datafactory/data_factory_pipeline_resource.go#L135-L139)
* [Looking at variable expansion a little deeper finds you](https://github.com/hashicorp/terraform-provider-azurerm/blob/d1f34cbc1754e4c29f9ab894bca1a89d431f0cfc/internal/services/datafactory/data_factory.go#L112-L123)
* [Even deeper, if you inspect the above of VariableSpecification](https://raw.githubusercontent.com/hashicorp/terraform-provider-azurerm/d1f34cbc1754e4c29f9ab894bca1a89d431f0cfc/vendor/github.com/Azure/azure-sdk-for-go/services/datafactory/mgmt/2018-06-01/datafactory/models.go)
 
```go
// VariableSpecification definition of a single variable for a Pipeline.
type VariableSpecification struct {
	// Type - Variable type. Possible values include: 'VariableTypeString', 'VariableTypeBool', 'VariableTypeArray'
	Type VariableType `json:"type,omitempty"`
	// DefaultValue - Default value of variable.
	DefaultValue interface{} `json:"defaultValue,omitempty"`
}
```

**Comment by embarrassedmilk** [comment]
Indeed, AzureSDK does support it, saw this code as well. Where I think it might get tricky is [here](https://github.com/hashicorp/terraform-provider-azurerm/blob/d1f34cbc1754e4c29f9ab894bca1a89d431f0cfc/internal/services/datafactory/data_factory_pipeline_resource.go#L61-L67). 
In ideal world i'd like to be able to do something like this:
```
variables = {
  var1 = "bar"
  var2 = false
  var3 = [1,2,3]
}
```
Also found [this issue](https://github.com/hashicorp/terraform-plugin-sdk/issues/616) suggesting that it's not possible at the moment.

**Comment by gkman** [comment]
Ya, I would suspect you'd have to just replace [this](https://github.com/hashicorp/terraform-provider-azurerm/blob/d1f34cbc1754e4c29f9ab894bca1a89d431f0cfc/internal/services/datafactory/data_factory_pipeline_resource.go#L65) with a validation function like [here[(https://github.com/hashicorp/terraform-provider-azurerm/blob/d1f34cbc1754e4c29f9ab894bca1a89d431f0cfc/internal/services/datafactory/data_factory_pipeline_resource.go#L98)

I'm not sure about the issue, I would have to dig into that. 

I would also be fine with something more complex like, but I think the more implicit the better in this case since that is the `terraform` way by default. 

```hcl
variables = {
  var1 = {
    type = string
    default = "point"  
  }
}
```

**Comment by embarrassedmilk** [comment]
Hmm, not sure if simple replacing will work. From the `Schema` [documentation](https://github.com/hashicorp/terraform-plugin-sdk/blob/v2.10.1/helper/schema/schema.go#L145-L147) it looks like it needs map elements to be one of the primitive types, so nothing complex is accepted.
The only type-safe workaround I can think of is to do something like [this](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/datafactory/data_factory_trigger_tumbling_window_resource.go#L174) which is also not ideal (in my opinion). 
Or we can turn `variables` into a json string, deserialize it and pass to the pipeline after it's validated. Something similar to [iam policy](https://github.com/hashicorp/terraform-provider-aws/blob/main/internal/service/iam/policy.go#L47-L52) in aws provider. Usage would look like this:
```
variables = jsonencode({
  var1 = {
    type = string
    default = "point"  
  })
}
```

**Comment by luisamador** [comment]
I've just come across this issue too and it's preventing my team from using Terraform for managing our Azure Data Factory configurations.

**Comment by paradox2707** [comment]
Please, fix this issue, I have needed pipeline variable with Array type and I had to find a workaround for it. In my case it was possible, but next time I doubt it will be affordable

**Comment by abratv** [comment]
https://github.com/hashicorp/terraform-provider-azurerm/issues/13131

**Comment by JuupHietbrink** [comment]
https://github.com/hashicorp/terraform-provider-azurerm/issues/13051

**Comment by katbyte** [comment]
It appears this issue covers the same ground as #13131 (Support for adding type to parameters and variables in azurerm_data_factory_pipeline), which is still open and has more of the discussion, so we are going to close this one out and keep it all in one place.

Please follow and thumbs up #13131 for updates. If these are actually different, please say so in a comment there or open a new issue referencing this one. Thank you!
